test/unittest/tests/block/test_block_api_write.cpp

```diff
@@ -0,0 +1,323 @@
+            REQUIRE(bm.write(&bm, session->get_wt_session_impl(), &buf, cookie.addr.data(),
+                      &cookie.size, false, false) == 0);
+
+            validate_write_block(&bm, session, &buf, cookie, str, false);
+            // Keep track of all the cookies, so that we can validate it again later.
+            cookies.push_back({std::move(cookie.addr), cookie.size});
+            __wt_buf_free(nullptr, &buf);
+        }
+        test_validate_cookies(&bm, session, cookies, test_strings);
+    }
+
+    SECTION("Test complex write api with changing write size")
+    {
+        std::vector<std::string> test_strings(
+          {"hello", std::move(std::string(300, 'a')), std::move(std::string(550, 'c')),
+            std::move(std::string(900, 'd')), std::move(std::string(1400, 'd'))});
+        std::vector<addr_cookie> cookies;
+        for (const auto &str : test_strings) {
+            WT_ITEM buf;
+            WT_CLEAR(buf);
+            test_and_validate_write_size(&bm, session, str.length());
+            create_write_buffer(&bm, session, str, &buf, str.length());
+
+            addr_cookie cookie;
+            REQUIRE(bm.write(&bm, session->get_wt_session_impl(), &buf, cookie.addr.data(),
+                      &cookie.size, false, false) == 0);
+
+            validate_write_block(&bm, session, &buf, cookie, str, false);
+            // Keep track of all the cookies, so that we can validate it again later.
+            cookies.push_back({std::move(cookie.addr), cookie.size});
+            __wt_buf_free(nullptr, &buf);
+        }
+        test_validate_cookies(&bm, session, cookies, test_strings);
+    }
+
+    SECTION("Test os_cache_dirty_max option")
+    {
+        // When this is set, when the file bytes written is greater an fsync should be called, and
+        // the file bytes written is set back to zero.
+        bm.block->os_cache_dirty_max = 500;
+
+        std::string test_string(200, 'a');
+        WT_ITEM buf;
+        WT_CLEAR(buf);
+        create_write_buffer(&bm, session, test_string, &buf, 0);
+
+        // The first block write should succeed.
+        addr_cookie cookie;
+        REQUIRE(bm.write(&bm, session->get_wt_session_impl(), &buf, cookie.addr.data(),
+                  &cookie.size, false, false) == 0);
+        validate_write_block(&bm, session, &buf, cookie, test_string, false);
+        REQUIRE(bm.block->fh->written == std::stoi(ALLOCATION_SIZE));
+
+        // At this point the file written is greater than os_cache_dirty_max, make sure that
+        // the session flag must be set before the fh->written is cleared.
+        REQUIRE(bm.write(&bm, session->get_wt_session_impl(), &buf, cookie.addr.data(),
+                  &cookie.size, false, false) == 0);
+        validate_write_block(&bm, session, &buf, cookie, test_string, false);
+        REQUIRE(bm.block->fh->written == std::stoi(ALLOCATION_SIZE) * 2);
+
+        // Flag is now set, the block write should flushed with fsync.
```

Comment:
I don't think this validates that we actually called fsync? But at the same time I doubt we can.

Reply:
yeah it was hard to test this, the only difference was fh->written member.